task/net/README.md

```diff
@@ -1,6 +1,155 @@
+
+`recv_packet` takes a socket identifier and a byte buffer as a lease. It asks
+the network stack to dequeue the next packet waiting on that socket and copy it
+into the leased buffer, returning metadata (source address and port). If no
+packet is waiting, it will return an error.
+
+Finally, whenever new activity occurs on a socket, the network stack will post
+the configured notification to the socket's owner. This means a server can
+service other requests and do useful work without sitting blocked on the
+netstack forever. When it notices the netstack's notification, it can turn
+around and talk to the netstack, which will return promptly without blocking.
+
+## Life of a UDP exchange / every byte copy
+
+This section describes every time a byte is transferred during a UDP
+receive-send pair. In total, not counting internal FIFOs within the Ethernet
+controller, each byte in a received UDP packet being routed to a valid socket
+associated with a task is written to memory three times. First, by the
+controller into a central DMA queue. Second, by the netstack into a
+socket-specific queue. Third, by the task that owns the socket, to transfer the
+data into its address space. The copy count for transmitted data is similar.
+
+We could potentially eliminate one of these copies by using a much fancier
+buffer pool management algorithm in the netstack, at the cost of complexity.
+
+Here is the current actual flow for a UDP request/response.
+
+1. Packet arrives from the Ethernet PHY into the MAC's receive FIFO.
+
+2. MAC notices packet is complete and verifies checksums, etc. It begins
+transferring the packet out of the FIFO and into a reserved area of memory using
+DMA.
+
+3. Once the DMA transfer is complete, the MAC generates an interrupt.
+
+4. The Hubris kernel handles the interrupt and notifies the `net` task.
+
+5. Once anything higher-priority has yielded the CPU, the `net` task gets the
+notification and inspects the packet in-place in DMA memory using `smoltcp`.
+Some packets, such as neighbor discovery and ICMP echo, are handled at this
+stage without a further copy. UDP packets are matched to a socket; if no
+matching socket exists, they are discarded with a Destination Port Unreachable
+response.
+
+6. If the matching socket has free buffer space, the packet is copied into the
+socket's buffer, and the DMA buffer is returned to the hardware to receive more
+packets. The `net` task notifies the task that owns the socket. If a packet is
+received for a socket that is out of space, the `net` task drops it, increments
+a counter on the socket, and returns the buffer to the DMA pool. (This copy is
+performed to ensure that a task that fails to read its socket in a timely
+fashion only stalls packets to _that socket,_ rather than starving the Ethernet
+DMA engine of buffers.)
+
+7. Once the `net` task and anything else higher priority has yielded, the owning
+task wakes up and sends a `recv_packet` message to `net`, loaning a writable
```

Comment:
maybe specify something like "wakes up (due to the notification)"?

Reply:
Yeah, and maybe something pointing out that this only happens if it was listening for it.